### PR TITLE
Migrate mul_mkldnn_op to phi matmul_kernel

### PR DESCRIPTION
**Pull request by HulekJakub**
### PR types
Others

### PR changes
 OPs

### Describe
Migrate mul_mkldnn_op to phi

### CODE REVIEW

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+void MatmulWithFlattenKernelINT8(const Context &dev_ctx,
+                                 const DenseTensor &x,
+                                 const DenseTensor &y,
+                                 int x_num_col_dims,
+                                 int y_num_col_dims,
+                                 DenseTensor *out) {
+  PADDLE_ENFORCE_EQ(paddle::platform::is_cpu_place(dev_ctx.GetPlace()),
+                    true,
+                    paddle::platform::errors::PreconditionNotMet(
+                        "Operator DNNL Mul must use CPUPlace"));
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
                    paddle::platform::errors::PreconditionNotMet(
                        "oneDNN MatmulWithFlatten kernel must use CPUPlace"));
```
now errors are located in `phi::errors` namespace

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        std::make_shared<MulPrimitiveFactory<XT, YT, OT>>(onednn_engine);
+    dev_ctx.SetBlob(key, prim_creator);
+  }
+
+  return prim_creator;
+}
+
+/* XT: input x data type, YT: input y data type */
+template <typename XT, typename YT, typename Context>
+inner_product_forward GetMulPrimitive(const Context &dev_ctx,
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
template <typename XT, typename YT>
inner_product_forward GetMulPrimitive(const Context &dev_ctx,
```
`OneDNNContext` is already passed as argument

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+
+  auto mul = GetMulPrimitive<XT, float, Context>(
+      dev_ctx, &x, &y, out, x_num_col_dims, y_num_col_dims, onednn_engine);
+
+  auto out_dims = out->dims();
+  if (out_dims.size() != 2) {
+    out->Resize(out_dims);
+  }
+
+  auto in_md = dnnl::memory::desc(*dnnl_primitive_desc_query_md(
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
  auto in_md = memory::desc(*dnnl_primitive_desc_query_md(
```

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+            ? PADDLE_GET_CONST(std::vector<float>, ctx.GetDnnAttr("scale_y"))
+            : std::vector<float>();
+    auto scale_x_data = ctx.HasDnnAttr("scale_x")
+                            ? PADDLE_GET_CONST(float, ctx.GetDnnAttr("scale_x"))
+                            : 1.0f;
+    auto scale_out_data =
+        force_fp32_output ? 1.0f
+        : ctx.HasDnnAttr("scale_out")
+            ? PADDLE_GET_CONST(float, ctx.GetDnnAttr("scale_out"))
+            : 1.0f;
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
  auto scale_out = ctx.HasDnnAttr("scale_out")
                       ? PADDLE_GET_CONST(float, ctx.GetDnnAttr("scale_out"))
                       : 1.0f;
  auto scale_out_data = force_fp32_output ? 1.0f : scale_out;
```
I think it's easier to read with two variables

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    auto user_y_desc =
+        CreateMemDescriptor<YT>(y_dims, funcs::OneDNNMemoryFormat::oi);
+    auto y_desc =
+        CreateMemDescriptor<int8_t>(y_dims, funcs::OneDNNMemoryFormat::oi);
+
+    return ReorderWithScale(
+        user_y_desc, y_desc, input_y.get_data_handle(), scale_y);
+  }
+
+  dnnl::primitive_attr CreateMulAttr(const OneDNNContext &ctx,
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
  dnnl::primitive_attr CreateMulAttr(const OneDNNContext &ctx,
```
Please rename each OneDNNContext `ctx` to `dev_ctx` to be consistent with already migrated oneDNN kernels

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+  }
+
+  memory CreateDstMemory(
+      const inner_product_forward::primitive_desc &mul_prim_desc,
+      const OneDNNContext &ctx,
+      DenseTensor *output) {
+    auto dst_desc = mul_prim_desc.dst_desc();
+    auto buffer_size = dst_desc.get_size();
+
+    OT *output_data = output->mutable_data<OT>(ctx.GetPlace(), buffer_size);
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
    OT *output_data = ctx.template Alloc<OT>(out, buffer_size);
```
`mutable_data` can't be used in PHI

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+    }
+
+    return data_matrix;
+  }
+
+  void UpdateDataPointers(const OneDNNContext &ctx,
+                          DenseTensor *out,
+                          const DenseTensor *in) {
+    x_input_->set_data_handle(phi::funcs::to_void_cast<XT>(in->data<XT>()));
+    output_->set_data_handle(out->mutable_data<OT>(ctx.GetPlace()));
```
Comment (Silv3S) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]:
```suggestion
    output_->set_data_handle(ctx.template Alloc<OT>(out));
```

**Review by Silv3S — DISMISSED** [review]
LGTM

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 8]
```diff
@@ -152,6 +161,420 @@ void MatmulKernel(const Context &dev_ctx,
...
+std::shared_ptr<MulPrimitiveFactory<XT, YT, OT>> GetPrimitiveFactory(
+    const OneDNNContext &dev_ctx,
+    const DenseTensor *input_x,
+    const DenseTensor *input_y,
+    const engine &onednn_engine) {
+  std::string key =
+      funcs::CreateKey(dev_ctx,
+                       paddle::framework::TransToProtoVarType(input_x->dtype()),
+                       phi::vectorize(input_x->dims()),
+                       paddle::framework::TransToProtoVarType(input_y->dtype()),
```
Comment (YuanRisheng) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]: We don't recommend using paddle::framework::TransToProtoVarType, you can try phi::TransToProtoVarType that is in phi/core/utils/data_type.h

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 9]
```diff
@@ -152,6 +161,420 @@ void MatmulKernel(const Context &dev_ctx,
...
+void MatmulWithFlattenKernelINT8(const Context &dev_ctx,
+                                 const DenseTensor &x,
+                                 const DenseTensor &y,
+                                 int x_num_col_dims,
+                                 int y_num_col_dims,
+                                 DenseTensor *out) {
+  PADDLE_ENFORCE_EQ(paddle::platform::is_cpu_place(dev_ctx.GetPlace()),
+                    true,
+                    paddle::platform::errors::PreconditionNotMet(
+                        "oneDNN MatmulWithFlatten kernel must use CPUPlace"));
```
Comment (YuanRisheng) [inline paddle/phi/kernels/onednn/matmul_kernel.cc:525]: paddle::platform::errors -> phi::errors

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 10]
```diff
@@ -152,6 +161,419 @@ void MatmulKernel(const Context &dev_ctx,
...
+                                : CreateMemDescriptor<T>(
+                                      data, funcs::OneDNNMemoryFormat::nchw))
+                         : src_mdesc;
+
+    if (src_mdesc != dst_mdesc) {
+      dev_ctx.template Alloc<T>(&x_tmp, data->memory_size());
+
+      Reorder(src_mdesc,
+              dst_mdesc,
+              phi::funcs::to_void_cast<T>(data->data<T>()),
```
Comment (paulinagacek) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]: you can get rid of `phi::` here as well as in other places in the code where method is in phi namespace

File: paddle/phi/kernels/onednn/matmul_kernel.cc [hunk 11]
```diff
@@ -152,6 +161,419 @@ void MatmulKernel(const Context &dev_ctx,
...
+
+/* XT: input x data type */
+template <typename XT, typename Context>
+void MatmulWithFlattenKernelINT8(const Context &dev_ctx,
+                                 const DenseTensor &x,
+                                 const DenseTensor &y,
+                                 int x_num_col_dims,
+                                 int y_num_col_dims,
+                                 DenseTensor *out) {
+  PADDLE_ENFORCE_EQ(paddle::platform::is_cpu_place(dev_ctx.GetPlace()),
```
Comment (paulinagacek) [inline paddle/phi/kernels/onednn/matmul_kernel.cc]: I think that `dev_ctx.GetPlace().GetType() == AllocationType::CPU` would be consistent with migrating instruction: 
"platform::is_xxx_place | place.GetType() == phi::AllocationType::XXX"

**Review by Silv3S — APPROVED** [review]

**Review by paulinagacek — APPROVED** [review]
LGTM!

**Review by YuanRisheng — APPROVED** [review]

**Pull request merged**